HOLD - CLOUDSTACK-10132: Multiple Management Servers Support for agents

nvazquez opened the pull request:
JIRA Ticket: https://issues.apache.org/jira/browse/CLOUDSTACK-10132

Functional Specification: https://cwiki.apache.org/confluence/display/CLOUDSTACK/Multiple+Management+Servers+Support+on+agents

This feature allows the direct agents to support multiple management servers. With this feature, a load balancer may no longer be necessary. The CloudStack administrator is responsible for setting a list of management servers and an algorithm in which to sort them, to the management server, using global configurations. The management server is responsible for applying algorithm to the management server list and propagate it to the agents. 

### Overview
A new global configuration is added: 'direct.agent.lb.algorithm'. Possible values of this global configuration are: 'static', 'roundrobin' and 'shuffle', corresponding to the algorithm that would be applied to the list of comma separated hosts defined on 'host' global configuration before sending them to the agents.
- 'static': No modification is made to the original 'host' list
- 'roundrobin': Rotate items one by one once the list is read
- 'shuffle': Randomly sort the list

A central utility is reading values from both global configurations on management server startup and is responsible of propagating list to each agent. In propagated list, the first element acts as the main host. Load balancing takes place by prepending the management server each agent should connect to, to the propagated actual list.

Once agents connect to management server, they receive a comma separated string of management servers. This value is compared to their actual 'host' configuration, and in case they differ, 'host' property is updated and persisted on 'agent.properties' file with the newly received string.

Reconnection logic between agent and management server is refactored as well. If agent is not able to establish a connection to its actual management server after some timeout interval, it attempts connection to the next element on its 'host' list.

Also, a background task is added to the agents to attempt reconnection to their main management server in pursuit of balancing. The interval of this background task is set on 'hosts.lb.background.interval' property on 'agent.properties' file. It can be optionally disabled, by setting property to 0

### Example
Supposing an environment in which there are 3 management servers: A, B and C and 3 KVM agents.
Setting 'host' = 'A,B,C', agents will receive lists depending on 'direct.agent.lb' value:
- 'static': Each agent will receive the list: 'A,B,C'
- 'roundrobin': First agent receives: 'A,B,C', second agent receives: 'B,C,A', third agent receives: 'C,B,A'
- 'shuffle': Each agent will receive a list in random order.

borisstoyanov commented:
@blueorangutan package

blueorangutan commented:
@borisstoyanov a Jenkins job has been kicked to build packages. I'll keep you posted as I make progress.

borisstoyanov commented:
@nvazquez thanks for this PR, I'll run marvin tests

blueorangutan commented:
Packaging result: ✔centos6 ✔centos7 ✔debian. JID-1214

borisstoyanov commented:
@blueorangutan test

blueorangutan commented:
@borisstoyanov a Trillian-Jenkins test job (centos7 mgmt + kvm-centos7) has been kicked to run smoke tests

rhtyd commented on agent/src/com/cloud/agent/AgentShell.java:
```diff
@@ -107,18 +109,37 @@ public String getPod() {
     @Override
-    public String getHost() {
+    public String getNextHost() {
         final String[] hosts = _host.split(",");
```
We can use getHosts() here.

rhtyd commented on agent/src/com/cloud/agent/AgentShell.java:
```diff
@@ -173,6 +194,19 @@ public void setPersistentProperty(String prefix, String name, String value) {
     }
 
+    @Override
+    public boolean setPropertyAndPersist(String property, String value) {
```
Don't use it, there is another way to do that --> call setPersistentProperty, see the method is just above your implementation.

rhtyd commented on utils/src/main/java/com/cloud/utils/PropertiesUtil.java:
```diff
@@ -193,4 +195,19 @@ public static Properties loadFromFile(final File file)
         return properties;
     }
 
+    public static void persist(Properties properties, File file, String key, String value) throws IOException {
```
This can be removed as well. See how the keystore passphrase is persisted, follow that pattern. Hope this helps @nvazquez

nvazquez commented:
Thanks @rhtyd for pointing out pattern for persistent properties, I've addressed your comments

borisstoyanov commented:
@blueorangutan package

blueorangutan commented:
@borisstoyanov a Jenkins job has been kicked to build packages. I'll keep you posted as I make progress.

blueorangutan commented:
Packaging result: ✖centos6 ✔centos7 ✖debian. JID-1221

borisstoyanov commented:
@blueorangutan package

blueorangutan commented:
@borisstoyanov a Jenkins job has been kicked to build packages. I'll keep you posted as I make progress.

blueorangutan commented:
Packaging result: ✔centos6 ✔centos7 ✔debian. JID-1222

borisstoyanov commented:
@blueorangutan package

blueorangutan commented:
@borisstoyanov a Jenkins job has been kicked to build packages. I'll keep you posted as I make progress.

blueorangutan commented:
Packaging result: ✔centos6 ✔centos7 ✔debian. JID-1225

rhtyd commented:
@blueorangutan package

blueorangutan commented:
@rhtyd a Jenkins job has been kicked to build packages. I'll keep you posted as I make progress.

blueorangutan commented:
Packaging result: ✔centos6 ✔centos7 ✔debian. JID-1229

rhtyd commented:
@nvazquez @borisstoyanov I've made and pushed some additional edge cases and bugfixes now. In my test env it seemed okay. Please continue reviewing and testing.
@blueorangutan package

blueorangutan commented:
@rhtyd a Jenkins job has been kicked to build packages. I'll keep you posted as I make progress.

blueorangutan commented:
Packaging result: ✖centos6 ✔centos7 ✖debian. JID-1238

rhtyd commented:
@blueorangutan package

blueorangutan commented:
@rhtyd a Jenkins job has been kicked to build packages. I'll keep you posted as I make progress.

blueorangutan commented:
Packaging result: ✔centos6 ✔centos7 ✔debian. JID-1240

rhtyd commented:
@blueorangutan test

blueorangutan commented:
@rhtyd a Trillian-Jenkins test job (centos7 mgmt + kvm-centos7) has been kicked to run smoke tests

blueorangutan commented:
<b>Trillian test result (tid-1641)</b>
Environment: kvm-centos7 (x2), Advanced Networking with Mgmt server 7
Total time taken: 40723 seconds
Marvin logs: https://github.com/blueorangutan/acs-prs/releases/download/trillian/pr2309-t1641-kvm-centos7.zip
Intermitten failure detected: /marvin/tests/smoke/test_deploy_virtio_scsi_vm.py
Intermitten failure detected: /marvin/tests/smoke/test_host_annotations.py
Intermitten failure detected: /marvin/tests/smoke/test_internal_lb.py
Intermitten failure detected: /marvin/tests/smoke/test_iso.py
Intermitten failure detected: /marvin/tests/smoke/test_privategw_acl.py
Intermitten failure detected: /marvin/tests/smoke/test_router_dnsservice.py
Intermitten failure detected: /marvin/tests/smoke/test_ssvm.py
Intermitten failure detected: /marvin/tests/smoke/test_vpc_redundant.py
Intermitten failure detected: /marvin/tests/smoke/test_vpc_vpn.py
Test completed. 56 look OK, 7 have error(s)


Test | Result | Time (s) | Test File
--- | --- | --- | ---
test_01_vpc_remote_access_vpn | `Failure` | 60.75 | test_vpc_vpn.py
test_08_reboot_cpvm | `Failure` | 0.03 | test_ssvm.py
test_06_stop_cpvm | `Failure` | 0.03 | test_ssvm.py
test_04_cpvm_internals | `Failure` | 0.03 | test_ssvm.py
test_02_list_cpvm_vm | `Failure` | 0.03 | test_ssvm.py
test_04_rvpc_privategw_static_routes | `Failure` | 262.55 | test_privategw_acl.py
test_03_vpc_privategw_restart_vpc_cleanup | `Failure` | 222.35 | test_privategw_acl.py
test_02_vpc_privategw_static_routes | `Failure` | 172.15 | test_privategw_acl.py
test_01_vpc_privategw_acl | `Failure` | 66.25 | test_privategw_acl.py
test_10_destroy_cpvm | `Error` | 5.12 | test_ssvm.py
test_03_create_iso_with_checksum_md5 | `Error` | 65.42 | test_iso.py
test_03_1_create_iso_with_checksum_md5_negative | `Error` | 60.57 | test_iso.py
test_02_create_iso_with_checksum_sha256 | `Error` | 65.45 | test_iso.py
test_01_create_iso_with_checksum_sha1 | `Error` | 65.40 | test_iso.py
test_01_1_create_iso_with_checksum_sha1_negative | `Error` | 40.48 | test_iso.py
ContextSuite context=TestInternalLb>:setup | `Error` | 0.00 | test_internal_lb.py
test_05_add_annotation_for_invalid_entityType | `Error` | 0.05 | test_host_annotations.py
ContextSuite context=TestDeployVirtioSCSIVM>:teardown | `Error` | 48.53 | test_deploy_virtio_scsi_vm.py
test_change_service_offering_for_vm_with_snapshots | Skipped | 0.00 | test_vm_snapshots.py
test_09_copy_delete_template | Skipped | 0.02 | test_templates.py
test_06_copy_template | Skipped | 0.00 | test_templates.py
test_static_role_account_acls | Skipped | 0.01 | test_staticroles.py
test_11_ss_nfs_version_on_ssvm | Skipped | 0.02 | test_ssvm.py
test_01_scale_vm | Skipped | 0.00 | test_scale_vm.py
test_01_primary_storage_iscsi | Skipped | 0.07 | test_primary_storage.py
test_vm_nic_adapter_vmxnet3 | Skipped | 0.00 | test_nic_adapter_type.py
test_03_nic_multiple_vmware | Skipped | 0.71 | test_nic.py
test_nested_virtualization_vmware | Skipped | 0.00 | test_nested_virtualization.py
test_06_copy_iso | Skipped | 0.00 | test_iso.py
test_list_ha_for_host_valid | Skipped | 0.02 | test_hostha_simulator.py
test_list_ha_for_host_invalid | Skipped | 0.02 | test_hostha_simulator.py
test_list_ha_for_host | Skipped | 0.02 | test_hostha_simulator.py
test_hostha_enable_feature_without_setting_provider | Skipped | 0.02 | test_hostha_simulator.py
test_hostha_enable_feature_valid | Skipped | 0.02 | test_hostha_simulator.py
test_hostha_disable_feature_valid | Skipped | 0.02 | test_hostha_simulator.py
test_hostha_configure_invalid_provider | Skipped | 0.02 | test_hostha_simulator.py
test_hostha_configure_default_driver | Skipped | 0.02 | test_hostha_simulator.py
test_ha_verify_fsm_recovering | Skipped | 0.02 | test_hostha_simulator.py
test_ha_verify_fsm_fenced | Skipped | 0.02 | test_hostha_simulator.py
test_ha_verify_fsm_degraded | Skipped | 0.03 | test_hostha_simulator.py
test_ha_verify_fsm_available | Skipped | 0.02 | test_hostha_simulator.py
test_ha_multiple_mgmt_server_ownership | Skipped | 0.02 | test_hostha_simulator.py
test_ha_list_providers | Skipped | 0.02 | test_hostha_simulator.py
test_ha_enable_feature_invalid | Skipped | 0.02 | test_hostha_simulator.py
test_ha_disable_feature_invalid | Skipped | 0.02 | test_hostha_simulator.py
test_ha_configure_enabledisable_across_clusterzones | Skipped | 0.02 | test_hostha_simulator.py
test_configure_ha_provider_valid | Skipped | 0.02 | test_hostha_simulator.py
test_configure_ha_provider_invalid | Skipped | 0.02 | test_hostha_simulator.py
test_deploy_vgpu_enabled_vm | Skipped | 0.02 | test_deploy_vgpu_enabled_vm.py
test_3d_gpu_support | Skipped | 0.03 | test_deploy_vgpu_enabled_vm.py

blueorangutan commented:
<b>Trillian test result (tid-1647)</b>
Environment: kvm-centos7 (x2), Advanced Networking with Mgmt server 7
Total time taken: 31123 seconds
Marvin logs: https://github.com/blueorangutan/acs-prs/releases/download/trillian/pr2309-t1647-kvm-centos7.zip
Intermitten failure detected: /marvin/tests/smoke/test_host_annotations.py
Intermitten failure detected: /marvin/tests/smoke/test_privategw_acl.py
Intermitten failure detected: /marvin/tests/smoke/test_vpc_redundant.py
Intermitten failure detected: /marvin/tests/smoke/test_vpc_vpn.py
Test completed. 60 look OK, 3 have error(s)


Test | Result | Time (s) | Test File
--- | --- | --- | ---
test_01_vpc_remote_access_vpn | `Failure` | 45.97 | test_vpc_vpn.py
test_04_rvpc_privategw_static_routes | `Failure` | 309.01 | test_privategw_acl.py
test_03_vpc_privategw_restart_vpc_cleanup | `Failure` | 188.37 | test_privategw_acl.py
test_02_vpc_privategw_static_routes | `Failure` | 107.90 | test_privategw_acl.py
test_01_vpc_privategw_acl | `Failure` | 61.78 | test_privategw_acl.py
test_05_add_annotation_for_invalid_entityType | `Error` | 0.10 | test_host_annotations.py
test_change_service_offering_for_vm_with_snapshots | Skipped | 0.00 | test_vm_snapshots.py
test_09_copy_delete_template | Skipped | 0.02 | test_templates.py
test_06_copy_template | Skipped | 0.00 | test_templates.py
test_static_role_account_acls | Skipped | 0.03 | test_staticroles.py
test_11_ss_nfs_version_on_ssvm | Skipped | 0.02 | test_ssvm.py
test_01_scale_vm | Skipped | 0.00 | test_scale_vm.py
test_01_primary_storage_iscsi | Skipped | 0.10 | test_primary_storage.py
test_vm_nic_adapter_vmxnet3 | Skipped | 0.00 | test_nic_adapter_type.py
test_03_nic_multiple_vmware | Skipped | 1.10 | test_nic.py
test_nested_virtualization_vmware | Skipped | 0.00 | test_nested_virtualization.py
test_06_copy_iso | Skipped | 0.00 | test_iso.py
test_list_ha_for_host_valid | Skipped | 0.03 | test_hostha_simulator.py
test_list_ha_for_host_invalid | Skipped | 0.03 | test_hostha_simulator.py
test_list_ha_for_host | Skipped | 0.03 | test_hostha_simulator.py
test_hostha_enable_feature_without_setting_provider | Skipped | 0.02 | test_hostha_simulator.py
test_hostha_enable_feature_valid | Skipped | 0.02 | test_hostha_simulator.py
test_hostha_disable_feature_valid | Skipped | 0.02 | test_hostha_simulator.py
test_hostha_configure_invalid_provider | Skipped | 0.03 | test_hostha_simulator.py
test_hostha_configure_default_driver | Skipped | 0.02 | test_hostha_simulator.py
test_ha_verify_fsm_recovering | Skipped | 0.02 | test_hostha_simulator.py
test_ha_verify_fsm_fenced | Skipped | 0.02 | test_hostha_simulator.py
test_ha_verify_fsm_degraded | Skipped | 0.03 | test_hostha_simulator.py
test_ha_verify_fsm_available | Skipped | 0.02 | test_hostha_simulator.py
test_ha_multiple_mgmt_server_ownership | Skipped | 0.02 | test_hostha_simulator.py
test_ha_list_providers | Skipped | 0.02 | test_hostha_simulator.py
test_ha_enable_feature_invalid | Skipped | 0.03 | test_hostha_simulator.py
test_ha_disable_feature_invalid | Skipped | 0.02 | test_hostha_simulator.py
test_ha_configure_enabledisable_across_clusterzones | Skipped | 0.03 | test_hostha_simulator.py
test_configure_ha_provider_valid | Skipped | 0.02 | test_hostha_simulator.py
test_configure_ha_provider_invalid | Skipped | 0.03 | test_hostha_simulator.py
test_deploy_vgpu_enabled_vm | Skipped | 0.05 | test_deploy_vgpu_enabled_vm.py
test_3d_gpu_support | Skipped | 0.04 | test_deploy_vgpu_enabled_vm.py

nvazquez commented:
@rhtyd @borisstoyanov please find new additions, kindly review/test

nvazquez commented:
@blueorangutan package

blueorangutan commented:
@nvazquez a Jenkins job has been kicked to build packages. I'll keep you posted as I make progress.

blueorangutan commented:
Packaging result: ✔centos6 ✔centos7 ✔debian. JID-1248

nvazquez commented:
@blueorangutan test

blueorangutan commented:
@nvazquez a Trillian-Jenkins test job (centos7 mgmt + kvm-centos7) has been kicked to run smoke tests

blueorangutan commented:
<b>Trillian test result (tid-1648)</b>
Environment: kvm-centos7 (x2), Advanced Networking with Mgmt server 7
Total time taken: 36300 seconds
Marvin logs: https://github.com/blueorangutan/acs-prs/releases/download/trillian/pr2309-t1648-kvm-centos7.zip
Intermitten failure detected: /marvin/tests/smoke/test_deploy_virtio_scsi_vm.py
Intermitten failure detected: /marvin/tests/smoke/test_host_annotations.py
Intermitten failure detected: /marvin/tests/smoke/test_privategw_acl.py
Intermitten failure detected: /marvin/tests/smoke/test_templates.py
Intermitten failure detected: /marvin/tests/smoke/test_vpc_redundant.py
Intermitten failure detected: /marvin/tests/smoke/test_vpc_vpn.py
Test completed. 58 look OK, 5 have error(s)


Test | Result | Time (s) | Test File
--- | --- | --- | ---
test_01_vpc_remote_access_vpn | `Failure` | 60.97 | test_vpc_vpn.py
test_04_rvpc_privategw_static_routes | `Failure` | 223.47 | test_privategw_acl.py
test_03_vpc_privategw_restart_vpc_cleanup | `Failure` | 157.94 | test_privategw_acl.py
test_02_vpc_privategw_static_routes | `Failure` | 152.91 | test_privategw_acl.py
test_01_vpc_privategw_acl | `Failure` | 61.72 | test_privategw_acl.py
ContextSuite context=TestVpcSite2SiteVpn>:setup | `Error` | 1597.80 | test_vpc_vpn.py
ContextSuite context=TestRVPCSite2SiteVpn>:setup | `Error` | 0.00 | test_vpc_vpn.py
ContextSuite context=TestVPCRedundancy>:setup | `Error` | 0.00 | test_vpc_redundant.py
test_05_add_annotation_for_invalid_entityType | `Error` | 0.07 | test_host_annotations.py
ContextSuite context=TestDeployVirtioSCSIVM>:teardown | `Error` | 45.56 | test_deploy_virtio_scsi_vm.py
test_change_service_offering_for_vm_with_snapshots | Skipped | 0.00 | test_vm_snapshots.py
test_09_copy_delete_template | Skipped | 0.02 | test_templates.py
test_06_copy_template | Skipped | 0.00 | test_templates.py
test_static_role_account_acls | Skipped | 0.02 | test_staticroles.py
test_11_ss_nfs_version_on_ssvm | Skipped | 0.02 | test_ssvm.py
test_01_scale_vm | Skipped | 0.00 | test_scale_vm.py
test_01_primary_storage_iscsi | Skipped | 0.09 | test_primary_storage.py
test_vm_nic_adapter_vmxnet3 | Skipped | 0.00 | test_nic_adapter_type.py
test_03_nic_multiple_vmware | Skipped | 1.07 | test_nic.py
test_nested_virtualization_vmware | Skipped | 0.00 | test_nested_virtualization.py
test_06_copy_iso | Skipped | 0.00 | test_iso.py
test_list_ha_for_host_valid | Skipped | 0.03 | test_hostha_simulator.py
test_list_ha_for_host_invalid | Skipped | 0.02 | test_hostha_simulator.py
test_list_ha_for_host | Skipped | 0.02 | test_hostha_simulator.py
test_hostha_enable_feature_without_setting_provider | Skipped | 0.02 | test_hostha_simulator.py
test_hostha_enable_feature_valid | Skipped | 0.02 | test_hostha_simulator.py
test_hostha_disable_feature_valid | Skipped | 0.02 | test_hostha_simulator.py
test_hostha_configure_invalid_provider | Skipped | 0.02 | test_hostha_simulator.py
test_hostha_configure_default_driver | Skipped | 0.02 | test_hostha_simulator.py
test_ha_verify_fsm_recovering | Skipped | 0.02 | test_hostha_simulator.py
test_ha_verify_fsm_fenced | Skipped | 0.02 | test_hostha_simulator.py
test_ha_verify_fsm_degraded | Skipped | 0.02 | test_hostha_simulator.py
test_ha_verify_fsm_available | Skipped | 0.02 | test_hostha_simulator.py
test_ha_multiple_mgmt_server_ownership | Skipped | 0.02 | test_hostha_simulator.py
test_ha_list_providers | Skipped | 0.02 | test_hostha_simulator.py
test_ha_enable_feature_invalid | Skipped | 0.02 | test_hostha_simulator.py
test_ha_disable_feature_invalid | Skipped | 0.02 | test_hostha_simulator.py
test_ha_configure_enabledisable_across_clusterzones | Skipped | 0.02 | test_hostha_simulator.py
test_configure_ha_provider_valid | Skipped | 0.02 | test_hostha_simulator.py
test_configure_ha_provider_invalid | Skipped | 0.03 | test_hostha_simulator.py
test_deploy_vgpu_enabled_vm | Skipped | 0.03 | test_deploy_vgpu_enabled_vm.py
test_3d_gpu_support | Skipped | 0.04 | test_deploy_vgpu_enabled_vm.py

rhtyd commented on agent/src/com/cloud/agent/Agent.java:
```diff
@@ -699,16 +719,97 @@ public void processResponse(final Response response, final Link link) {
+        if (msList != null && msList.size() > 0) {
+            final List<String> hosts = Arrays.asList(_shell.getHosts());
+            s_logger.info("Received management servers list: " + msList + ", current list: " + hosts);
+            if (isReceivedListUpdated(hosts, msList)) {
```
It's debatable if we want to update the `host` list when order has changed, I think if the hosts are same (but order is different) we should ignore. In my testing I was able to reproduce an infinite switching behaviour when preferred host (the first in list) was not the one the agent was connected to leading to period disconnections (with just two hosts in the list).

DaanHoogland replied:
That seems to be a design issue then. When the order has changed the (ordered) list should be updated. It is debatable if the connection should be reset, though.

rhtyd commented on agent/src/com/cloud/agent/Agent.java:
```diff
@@ -699,16 +719,97 @@ public void processResponse(final Response response, final Link link) {
+            if (isReceivedListUpdated(hosts, msList)) {
+                persistNewManagementServersList(msList);
+                if (!_shell.getConnectedHost().equals(msList.get(0))) {
+                    attemptReconnectionToPrimaryHost(msList.get(0));
```
Please remove the attempt to reconnect from here, the responsibiilty of doing that lies with the background task and not the ready command handler.

rhtyd commented on agent/src/com/cloud/agent/Agent.java:
```diff
@@ -990,4 +1091,53 @@ public void doTask(final Task task) throws TaskExecutionException {
+                 final long wait = _shell.getFailbackPollingWait();
+                 while (!preferredHost.equals(connectedHost)) {
+                     if (_link != null) {
+                         boolean isHostUp = true;
```
Optional: The code to check if the host is reachable on a port/address can we refactored.

rhtyd commented on agent/src/com/cloud/agent/Agent.java:
```diff
@@ -990,4 +1091,53 @@ public void doTask(final Task task) throws TaskExecutionException {
+                     s_logger.trace("Running host LB checker task, connected host=" + connectedHost + ", preferred host=" + preferredHost);
+                 }
+                 final long wait = _shell.getFailbackPollingWait();
+                 while (!preferredHost.equals(connectedHost)) {
```
Let's not have a while loop here. @nvazquez if you're not aware, when the agent is disconnected the `reconnect()` logic already kicks and is responsible for performing reconnections. This timer/task should not share the same responsibility, instead it should check if the preferred host (first in the list) is available and if the agent is already connected, only then it should disconnect by performing a reconnect on the `_link` (if it's not null for example).

Bear in mind, we've two assumptions here (1) the first host in the list is the preferred one, (2) the timer task is responsible for reconnecting to the preferred host.

nvazquez replied:
Done, thanks

rhtyd commented on agent/src/com/cloud/agent/Agent.java:
```diff
@@ -990,4 +1091,53 @@ public void doTask(final Task task) throws TaskExecutionException {
+                             }
+                         }
+                     }
+                     index = (index + 1) % msList.length;
```
@nvazquez don't change the preferred host to some other host, it's because we assume the first in the list is the `preferred` host. For example, consider this use-case when using the `static` algorithm the first item in the list could be IP of the LB (netscalar/haproxy etc) and the use of this mechanism is to ensure HA for the mgmt server when this LB is down. If we change the preferred host, it breaks our use-case.

nvazquez replied:
Thanks @rhtyd, I assumed background task should attempt reconnection to any available host starting from the preferred one at the beggining of the list and trying in order on 'host' list. I refactored this background task to attempt reconnection only to preferred host as you suggested, I think it simplifies the feature

rhtyd commented on agent/src/com/cloud/agent/Agent.java:
```diff
@@ -990,4 +1091,53 @@ public void doTask(final Task task) throws TaskExecutionException {
+                         }
+                         if (isHostUp && _link != null) {
+                             long initialTime = System.currentTimeMillis();
+                             while (System.currentTimeMillis() - initialTime < wait && _inProgress.get() > 0) {
```
@nvazquez please avoid writing procedural code, in `reconnect` remaining tasks handling already exists. We should not have any timeouts/limits on these tasks, i.e. don't kill/purge outstanding tasks. Instead, one way to rewrite this could be to perform no operations when there are any tasks in `_inProgress`, and let the HostLBCheckerTask retry in the next round.

nvazquez replied:
Done, thanks! Agreed

rhtyd commented on agent/conf/agent.properties:
```diff
@@ -30,6 +30,19 @@ workers=5
+# The time in seconds in which the agent will wait for active commands to be completed
+# before connecting to preferred host
+# Any change in this setting requires restarting the agent.
+failback.polling.wait=10
```
@nvazquez this can be removed, see my comments below.

rhtyd commented on agent/src/com/cloud/agent/Agent.java:
```diff
@@ -699,16 +719,97 @@ public void processResponse(final Response response, final Link link) {
+     * Try reconnecting to preferred management server after new list is received.
+     * If connection is not performed, host counter is reset
+     */
+    private void attemptReconnectionToPrimaryHost(String preferred) {
```
This can be removed, duplicates code too.

rhtyd commented on agent/src/com/cloud/agent/Agent.java:
```diff
@@ -699,16 +719,97 @@ public void processResponse(final Response response, final Link link) {
+        if (actual.size() != received.size()) {
+            return true;
+        }
+        for (int i = 0; i < received.size(); i++) {
```
We can simply using the previous implementation to simply check if the existing setting differs in size or `set`.

rhtyd commented on agent/src/com/cloud/agent/Agent.java:
```diff
@@ -990,4 +1091,53 @@ public void doTask(final Task task) throws TaskExecutionException {
+                                     s_logger.debug("Preferred host is found to be reachable, trying to reconnect to: " + preferredHost);
+                                 }
+                                 _shell.resetHostCounter();
+                                 reconnect(_link);
```
@nvazquez just so we're on the same page, when performing a reconnect after resting the host counter, `reconnect` will start getting host from the first in the list, i.e. the preferred one.

nvazquez replied:
Right, my bad. Now it makes sense as timer task will attempt reconnection to the first host in the list

rhtyd commented on agent/src/com/cloud/agent/AgentShell.java:
```diff
@@ -107,18 +111,48 @@ public String getPod() {
+        connectedHost = hosts[_hostCounter % hosts.length];
         _hostCounter++;
-        return host;
+        s_logger.info("Connecting to host: " + connectedHost);
```
This method only returns the next possible host in list, however, this does not mean we're connected to the host yet. This might cause threading issues, and in rare cases the connectedHost might not be the connected host. Reconsider how this could be re-written, perhaps the connectedHost can be updated using one of the `Task.CONNECT` handlers.

nvazquez replied:
Done, thanks! Please checkout `hostToConnect` and `connectedHost` relationship using `updateConnectedHost()` method on connection to a host.

rhtyd commented on agent/src/com/cloud/agent/AgentShell.java:
```diff
@@ -291,6 +325,11 @@ protected boolean parseCommand(final String[] args) throws ConfigurationExceptio
 
+        String val = getProperty(null, "host.lb.interval");
+        hostLBTimerInterval = (Strings.isNullOrEmpty(val) ? 0L : Long.valueOf(val));
+        String wait = getProperty(null, "failback.polling.wait");
```
Remove this.

rhtyd commented on agent/src/com/cloud/agent/IAgentShell.java:
```diff
@@ -22,33 +22,64 @@
+    /**
+     * Resets host counter to start from the beginning
+     */
+    void resetHostCounter();
```
Remove comments, or add them for all the methods on this interface for consistency.

DaanHoogland replied:
I do not feel consistency is an issue here as much as the comment not giving more information than the methodname, so yes, please remove.

rhtyd commented on server/src/com/cloud/resource/ResourceManagerImpl.java:
```diff
@@ -800,6 +804,7 @@ public Discoverer getMatchingDiscover(final Hypervisor.HypervisorType hypervisor
 
                 }
                 s_logger.info("server resources successfully discovered by " + discoverer.getName());
+                agentMSLB.propagateManagementServerListToAgents();
```
@nvazquez this is not desirable, this will cause the list to be sent to all the hosts whenever any single agent connects.

rhtyd commented on server/src/org/apache/cloudstack/agent/mslb/AgentMSLBServiceImpl.java:
```diff
@@ -0,0 +1,137 @@
+    public void propagateManagementServerListToAgents() {
+        List<HostVO> hosts = getOrderedRunningIndirectAgentIds();
+        if (hosts != null) {
+            for (HostVO h : hosts) {
```
@nvazquez I don't think it's a good idea, `resource claim and ownership` is not within the scope of this feature. This method will send command/answers to all the hosts (even those not managed by the `this` management server) and may cause disconnections. In case of multiple management servers, this will cause all the management servers to send multiple requests to all the agents. (note, when you do agent.send(), if the agent is not connected to `this` management server, it will use clustered mgmt server impl to send the request to the forwarding agent which is via other mgmt servers).

nvazquez commented:
@rhtyd thanks for your valuable review! I refactored code addressing most of your comments. There are quite few in which I think we should get aligned on and once we get aligned I could refactor as well! Thanks!

borisstoyanov commented:
@blueorangutan package

blueorangutan commented:
@borisstoyanov a Jenkins job has been kicked to build packages. I'll keep you posted as I make progress.

blueorangutan commented:
Packaging result: ✔centos6 ✔centos7 ✔debian. JID-1251

borisstoyanov commented:
@blueorangutan package

blueorangutan commented:
@borisstoyanov a Jenkins job has been kicked to build packages. I'll keep you posted as I make progress.

blueorangutan commented:
Packaging result: ✔centos6 ✔centos7 ✔debian. JID-1253

rhtyd commented on agent/src/com/cloud/agent/Agent.java:
```diff
@@ -138,12 +143,14 @@ public int value() {
     private String _keystoreSetupPath;
     private String _keystoreCertImportPath;
 
+    private Long connectionTimeout = 120000L;
```
@nvazquez This is not used anywhere, can be removed.

rhtyd commented on agent/src/com/cloud/agent/Agent.java:
```diff
@@ -216,6 +223,7 @@ public String getPod() {
     }
 
     protected void setLink(final Link link) {
+        s_logger.info("Setting up link!");
```
I had put this for logging/debugging, may be removed. @nvazquez

rhtyd commented on agent/src/com/cloud/agent/Agent.java:
```diff
@@ -265,6 +273,8 @@ public void start() {
                 s_logger.info("Attempted to connect to the server, but received an unexpected exception, trying again...");
             }
         }
+        _shell.updateConnectedHost();
```
@nvazquez to be on safer/deterministic side, can this method accept the `host` String and set that instead of `hostToConnect`?

nvazquez replied:
@rhtyd I decided to take that path as when agent is launched, the next host is set on constructor and then from `start()` method we couldn't know which host the agent is trying to connect to unless we do the same as in AgentShell with `hostToConnect` but on Agent class. Which approach do you think it would be best?

rhtyd commented on agent/src/com/cloud/agent/Agent.java:
```diff
@@ -699,16 +721,102 @@ public void processResponse(final Response response, final Link link) {
+                persistNewManagementServersList(msList);
+                _shell.resetHostCounter();
+                if (hasPrimaryHostChanged(hosts, msList)) {
+                    attemptReconnectionToPrimaryHost();
```
@nvazquez this can be removed, the ReadyCommand handler should not be responsible for reconnection/switching.

rhtyd commented on agent/src/com/cloud/agent/Agent.java:
```diff
@@ -699,16 +721,102 @@ public void processResponse(final Response response, final Link link) {
+     * Try reconnecting to preferred management server after new list is received.
+     * If connection is not performed, host counter is reset
+     */
+    private void attemptReconnectionToPrimaryHost() {
```
@nvazquez this method duplicates code and is not necessary, may be removed.

rhtyd commented on server/src/org/apache/cloudstack/agent/mslb/AgentMSLBServiceImpl.java:
```diff
@@ -0,0 +1,128 @@
+        List<HostVO> hosts = resourceManager.listAllHostsInOneZoneByType(Host.Type.Routing, dcId);
+        if (hosts != null) {
+            return hosts.stream()
+                    .filter(x -> x.getHypervisorType().equals(Hypervisor.HypervisorType.KVM) &&
```
This should not be restricted just for KVM, but for any connected (indirect) agent.

nvazquez commented:
@rhtyd listeners for global settings update and new hosts added. Also shuffle is back to "normal" behavior. Lists will be propagated to agents only when 'hosts' or 'connected.agent.mslb.algorithm' values change or new hosts added, using new command-answer pattern instead of `ReadyCommand`. Please kindly review changes

borisstoyanov commented:
@blueorangutan package

blueorangutan commented:
@borisstoyanov a Jenkins job has been kicked to build packages. I'll keep you posted as I make progress.

blueorangutan commented:
Packaging result: ✔centos6 ✔centos7 ✔debian. JID-1693

borisstoyanov commented:
@blueorangutan test

blueorangutan commented:
@borisstoyanov a Trillian-Jenkins test job (centos7 mgmt + kvm-centos7) has been kicked to run smoke tests

blueorangutan commented:
<b>Trillian test result (tid-2233)</b>
Environment: kvm-centos7 (x2), Advanced Networking with Mgmt server 7
Total time taken: 17493 seconds
Marvin logs: https://github.com/blueorangutan/acs-prs/releases/download/trillian/pr2309-t2233-kvm-centos7.zip
Intermitten failure detected: /marvin/tests/smoke/test_accounts.py
Intermitten failure detected: /marvin/tests/smoke/test_affinity_groups_projects.py
Intermitten failure detected: /marvin/tests/smoke/test_affinity_groups.py
Intermitten failure detected: /marvin/tests/smoke/test_deploy_virtio_scsi_vm.py
Intermitten failure detected: /marvin/tests/smoke/test_deploy_vm_iso.py
Intermitten failure detected: /marvin/tests/smoke/test_deploy_vm_root_resize.py
Intermitten failure detected: /marvin/tests/smoke/test_deploy_vms_with_varied_deploymentplanners.py
Intermitten failure detected: /marvin/tests/smoke/test_deploy_vm_with_userdata.py
Intermitten failure detected: /marvin/tests/smoke/test_internal_lb.py
Intermitten failure detected: /marvin/tests/smoke/test_iso.py
Intermitten failure detected: /marvin/tests/smoke/test_list_ids_parameter.py
Intermitten failure detected: /marvin/tests/smoke/test_loadbalance.py
Intermitten failure detected: /marvin/tests/smoke/test_metrics_api.py
Intermitten failure detected: /marvin/tests/smoke/test_migration.py
Intermitten failure detected: /marvin/tests/smoke/test_multipleips_per_nic.py
Intermitten failure detected: /marvin/tests/smoke/test_nested_virtualization.py
Intermitten failure detected: /marvin/tests/smoke/test_network_acl.py
Intermitten failure detected: /marvin/tests/smoke/test_network.py
Intermitten failure detected: /marvin/tests/smoke/test_nic.py
Intermitten failure detected: /marvin/tests/smoke/test_over_provisioning.py
Intermitten failure detected: /marvin/tests/smoke/test_password_server.py
Intermitten failure detected: /marvin/tests/smoke/test_portforwardingrules.py
Intermitten failure detected: /marvin/tests/smoke/test_primary_storage.py
Intermitten failure detected: /marvin/tests/smoke/test_privategw_acl.py
Intermitten failure detected: /marvin/tests/smoke/test_projects.py
Intermitten failure detected: /marvin/tests/smoke/test_public_ip_range.py
Intermitten failure detected: /marvin/tests/smoke/test_reset_vm_on_reboot.py
Intermitten failure detected: /marvin/tests/smoke/test_router_dhcphosts.py
Intermitten failure detected: /marvin/tests/smoke/test_router_dns.py
Intermitten failure detected: /marvin/tests/smoke/test_router_dnsservice.py
Intermitten failure detected: /marvin/tests/smoke/test_routers_iptables_default_policy.py
Intermitten failure detected: /marvin/tests/smoke/test_routers_network_ops.py
Intermitten failure detected: /marvin/tests/smoke/test_routers.py
Intermitten failure detected: /marvin/tests/smoke/test_secondary_storage.py
Intermitten failure detected: /marvin/tests/smoke/test_service_offerings.py
Intermitten failure detected: /marvin/tests/smoke/test_snapshots.py
Intermitten failure detected: /marvin/tests/smoke/test_ssvm.py
Intermitten failure detected: /marvin/tests/smoke/test_templates.py
Intermitten failure detected: /marvin/tests/smoke/test_usage.py
Intermitten failure detected: /marvin/tests/smoke/test_vm_life_cycle.py
Intermitten failure detected: /marvin/tests/smoke/test_vm_snapshots.py
Intermitten failure detected: /marvin/tests/smoke/test_volumes.py
Intermitten failure detected: /marvin/tests/smoke/test_vpc_redundant.py
Intermitten failure detected: /marvin/tests/smoke/test_vpc_router_nics.py
Intermitten failure detected: /marvin/tests/smoke/test_vpc_vpn.py
Intermitten failure detected: /marvin/tests/smoke/test_host_maintenance.py
Smoke tests completed. 22 look OK, 45 have error(s)
Only failed tests results shown below:


Test | Result | Time (s) | Test File
--- | --- | --- | ---
test_01_create_iso_with_checksum_sha1 | `Error` | 65.44 | test_iso.py
test_02_create_iso_with_checksum_sha256 | `Error` | 65.44 | test_iso.py
test_03_create_iso_with_checksum_md5 | `Error` | 65.45 | test_iso.py
test_04_create_iso_with_no_checksum | `Error` | 65.45 | test_iso.py
test_01_create_iso | `Failure` | 1514.07 | test_iso.py
ContextSuite context=TestISO>:setup | `Error` | 3029.37 | test_iso.py
ContextSuite context=TestAccounts>:setup | `Error` | 0.00 | test_accounts.py
ContextSuite context=TestAddVmToSubDomain>:setup | `Error` | 0.00 | test_accounts.py
test_DeleteDomain | `Error` | 1.39 | test_accounts.py
test_forceDeleteDomain | `Error` | 1.42 | test_accounts.py
ContextSuite context=TestRemoveUserFromAccount>:setup | `Error` | 5.71 | test_accounts.py
ContextSuite context=TestTemplateHierarchy>:setup | `Error` | 5.84 | test_accounts.py
ContextSuite context=TestInternalLb>:setup | `Error` | 0.00 | test_internal_lb.py
ContextSuite context=TestDeployVmWithAffinityGroup>:setup | `Error` | 0.00 | test_affinity_groups_projects.py
test_list_infrastructure_metrics | `Error` | 0.26 | test_metrics_api.py
test_list_pstorage_metrics | `Error` | 0.18 | test_metrics_api.py
test_list_vms_metrics | `Error` | 0.15 | test_metrics_api.py
ContextSuite context=TestDeployVmWithAffinityGroup>:setup | `Error` | 0.00 | test_affinity_groups.py
ContextSuite context=TestNetworkMigration>:setup | `Error` | 0.00 | test_migration.py
ContextSuite context=TestDeployVirtioSCSIVM>:setup | `Error` | 0.00 | test_deploy_virtio_scsi_vm.py
test_deploy_vm_from_iso | `Error` | 0.00 | test_deploy_vm_iso.py
ContextSuite context=TestNetworkACL>:setup | `Error` | 0.00 | test_network_acl.py
ContextSuite context=TestDeployVmRootSize>:setup | `Error` | 0.00 | test_deploy_vm_root_resize.py
ContextSuite context=TestDeployVmWithVariedPlanners>:setup | `Error` | 0.00 | test_deploy_vms_with_varied_deploymentplanners.py
test_delete_account | `Error` | 0.81 | test_network.py
test_delete_network_while_vm_on_it | `Error` | 0.08 | test_network.py
test_deploy_vm_l2network | `Error` | 0.07 | test_network.py
test_l2network_restart | `Error` | 1.18 | test_network.py
ContextSuite context=TestPortForwarding>:setup | `Error` | 3.47 | test_network.py
test_reboot_router | `Error` | 0.10 | test_network.py
test_releaseIP | `Error` | 0.76 | test_network.py
ContextSuite context=TestRouterRules>:setup | `Error` | 0.85 | test_network.py
ContextSuite context=TestDeployVmWithUserData>:setup | `Error` | 0.00 | test_deploy_vm_with_userdata.py
test_01_nic | `Error` | 0.08 | test_nic.py
ContextSuite context=TestLoadBalance>:setup | `Error` | 0.00 | test_loadbalance.py
test_UpdateStorageOverProvisioningFactor | `Failure` | 0.02 | test_over_provisioning.py
test_UpdateStorageOverProvisioningFactor | `Error` | 0.02 | test_over_provisioning.py
ContextSuite context=TestListIdsParams>:setup | `Error` | 0.00 | test_list_ids_parameter.py
test_nic_secondaryip_add_remove | `Error` | 0.09 | test_multipleips_per_nic.py
ContextSuite context=TestNestedVirtualization>:setup | `Error` | 0.00 | test_nested_virtualization.py
ContextSuite context=TestIsolatedNetworksPasswdServer>:setup | `Error` | 0.00 | test_password_server.py
ContextSuite context=TestPortForwardingRules>:setup | `Error` | 0.00 | test_portforwardingrules.py
test_01_add_primary_storage_disabled_host | `Error` | 32.65 | test_primary_storage.py
ContextSuite context=TestStorageTags>:setup | `Error` | 31.86 | test_primary_storage.py
ContextSuite context=TestPrivateGwACL>:setup | `Error` | 0.00 | test_privategw_acl.py
test_09_project_suspend | `Error` | 1.12 | test_projects.py
test_10_project_activation | `Error` | 1.11 | test_projects.py
test_02_create_template_with_checksum_sha1 | `Error` | 65.51 | test_templates.py
test_03_create_template_with_checksum_sha256 | `Error` | 65.49 | test_templates.py
test_04_create_template_with_checksum_md5 | `Error` | 65.50 | test_templates.py
test_05_create_template_with_no_checksum | `Error` | 65.49 | test_templates.py
test_02_deploy_vm_from_direct_download_template | `Error` | 0.11 | test_templates.py
test_03_deploy_vm_wrong_checksum | `Error` | 0.25 | test_templates.py
ContextSuite context=TestTemplates>:setup | `Error` | 3.62 | test_templates.py
ContextSuite context=TestResetVmOnReboot>:setup | `Error` | 0.00 | test_reset_vm_on_reboot.py
ContextSuite context=TestRouterDHCPHosts>:setup | `Error` | 0.00 | test_router_dhcphosts.py
ContextSuite context=TestRouterDHCPOpts>:setup | `Error` | 0.00 | test_router_dhcphosts.py
ContextSuite context=TestRouterDns>:setup | `Error` | 0.00 | test_router_dns.py
ContextSuite context=TestRouterDnsService>:setup | `Error` | 0.00 | test_router_dnsservice.py
ContextSuite context=TestRouterIpTablesPolicies>:setup | `Error` | 0.00 | test_routers_iptables_default_policy.py
ContextSuite context=TestVPCIpTablesPolicies>:setup | `Error` | 0.00 | test_routers_iptables_default_policy.py
test_01_isolate_network_FW_PF_default_routes_egress_true | `Error` | 0.18 | test_routers_network_ops.py
test_02_isolate_network_FW_PF_default_routes_egress_false | `Error` | 0.16 | test_routers_network_ops.py
ContextSuite context=TestRedundantIsolateNetworks>:setup | `Error` | 1.40 | test_routers_network_ops.py
ContextSuite context=TestISOUsage>:setup | `Error` | 0.00 | test_usage.py
ContextSuite context=TestLBRuleUsage>:setup | `Error` | 0.00 | test_usage.py
ContextSuite context=TestNatRuleUsage>:setup | `Error` | 0.00 | test_usage.py
ContextSuite context=TestPublicIPUsage>:setup | `Error` | 0.00 | test_usage.py
ContextSuite context=TestSnapshotUsage>:setup | `Error` | 0.00 | test_usage.py
ContextSuite context=TestVmUsage>:setup | `Error` | 0.00 | test_usage.py
ContextSuite context=TestVolumeUsage>:setup | `Error` | 0.00 | test_usage.py
ContextSuite context=TestVpnUsage>:setup | `Error` | 0.00 | test_usage.py
ContextSuite context=TestRouterServices>:setup | `Error` | 0.00 | test_routers.py
test_01_sys_vm_start | `Failure` | 0.11 | test_secondary_storage.py
test_02_sys_template_ready | `Failure` | 0.12 | test_secondary_storage.py
ContextSuite context=TestServiceOfferings>:setup | `Error` | 0.22 | test_service_offerings.py
ContextSuite context=TestSnapshotRootDisk>:setup | `Error` | 0.00 | test_snapshots.py
test_01_list_sec_storage_vm | `Failure` | 0.04 | test_ssvm.py
test_02_list_cpvm_vm | `Failure` | 0.04 | test_ssvm.py
test_03_ssvm_internals | `Failure` | 0.04 | test_ssvm.py
test_04_cpvm_internals | `Failure` | 0.04 | test_ssvm.py
test_05_stop_ssvm | `Failure` | 0.04 | test_ssvm.py
test_06_stop_cpvm | `Failure` | 0.04 | test_ssvm.py
test_07_reboot_ssvm | `Failure` | 0.04 | test_ssvm.py
test_08_reboot_cpvm | `Failure` | 0.04 | test_ssvm.py
test_09_destroy_ssvm | `Failure` | 0.04 | test_ssvm.py
test_10_destroy_cpvm | `Failure` | 0.04 | test_ssvm.py
ContextSuite context=TestDeployVM>:setup | `Error` | 0.00 | test_vm_life_cycle.py
ContextSuite context=TestVMLifeCycle>:setup | `Error` | 0.00 | test_vm_life_cycle.py
ContextSuite context=TestVmSnapshot>:setup | `Error` | 0.09 | test_vm_snapshots.py
ContextSuite context=TestCreateVolume>:setup | `Error` | 0.00 | test_volumes.py
ContextSuite context=TestVolumes>:setup | `Error` | 0.00 | test_volumes.py
test_01_create_redundant_VPC_2tiers_4VMs_4IPs_4PF_ACL | `Error` | 7.07 | test_vpc_redundant.py
test_02_redundant_VPC_default_routes | `Error` | 8.10 | test_vpc_redundant.py
test_03_create_redundant_VPC_1tier_2VMs_2IPs_2PF_ACL_reboot_routers | `Error` | 8.10 | test_vpc_redundant.py
test_04_rvpc_network_garbage_collector_nics | `Error` | 8.11 | test_vpc_redundant.py
test_05_rvpc_multi_tiers | `Error` | 8.10 | test_vpc_redundant.py
test_01_VPC_nics_after_destroy | `Error` | 6.03 | test_vpc_router_nics.py
test_02_VPC_default_routes | `Error` | 6.02 | test_vpc_router_nics.py
test_01_redundant_vpc_site2site_vpn | `Failure` | 6.48 | test_vpc_vpn.py
test_01_vpc_site2site_vpn_multiple_options | `Failure` | 4.38 | test_vpc_vpn.py
test_01_vpc_remote_access_vpn | `Failure` | 3.23 | test_vpc_vpn.py
test_01_vpc_site2site_vpn | `Failure` | 4.38 | test_vpc_vpn.py
test_01_cancel_host_maintenace_with_no_migration_jobs | `Failure` | 5.37 | test_host_maintenance.py
test_02_cancel_host_maintenace_with_migration_jobs | `Error` | 0.33 | test_host_maintenance.py

blueorangutan commented:
<b>Trillian test result (tid-2236)</b>
Environment: kvm-centos7 (x2), Advanced Networking with Mgmt server 7
Total time taken: 43 seconds
Marvin logs: https://github.com/blueorangutan/acs-prs/releases/download/trillian/pr2309-t2236-kvm-centos7.zip
Smoke tests completed. 1 look OK, 0 have error(s)
Only failed tests results shown below:


Test | Result | Time (s) | Test File
--- | --- | --- | ---

blueorangutan commented:
<b>Trillian test result (tid-2235)</b>
Environment: kvm-centos7 (x2), Advanced Networking with Mgmt server 7
Total time taken: 17565 seconds
Marvin logs: https://github.com/blueorangutan/acs-prs/releases/download/trillian/pr2309-t2235-kvm-centos7.zip
Intermitten failure detected: /marvin/tests/smoke/test_accounts.py
Intermitten failure detected: /marvin/tests/smoke/test_affinity_groups_projects.py
Intermitten failure detected: /marvin/tests/smoke/test_affinity_groups.py
Intermitten failure detected: /marvin/tests/smoke/test_deploy_virtio_scsi_vm.py
Intermitten failure detected: /marvin/tests/smoke/test_deploy_vm_iso.py
Intermitten failure detected: /marvin/tests/smoke/test_deploy_vm_root_resize.py
Intermitten failure detected: /marvin/tests/smoke/test_deploy_vms_with_varied_deploymentplanners.py
Intermitten failure detected: /marvin/tests/smoke/test_deploy_vm_with_userdata.py
Intermitten failure detected: /marvin/tests/smoke/test_internal_lb.py
Intermitten failure detected: /marvin/tests/smoke/test_iso.py
Intermitten failure detected: /marvin/tests/smoke/test_list_ids_parameter.py
Intermitten failure detected: /marvin/tests/smoke/test_loadbalance.py
Intermitten failure detected: /marvin/tests/smoke/test_metrics_api.py
Intermitten failure detected: /marvin/tests/smoke/test_migration.py
Intermitten failure detected: /marvin/tests/smoke/test_multipleips_per_nic.py
Intermitten failure detected: /marvin/tests/smoke/test_nested_virtualization.py
Intermitten failure detected: /marvin/tests/smoke/test_network_acl.py
Intermitten failure detected: /marvin/tests/smoke/test_network.py
Intermitten failure detected: /marvin/tests/smoke/test_nic.py
Intermitten failure detected: /marvin/tests/smoke/test_over_provisioning.py
Intermitten failure detected: /marvin/tests/smoke/test_password_server.py
Intermitten failure detected: /marvin/tests/smoke/test_portforwardingrules.py
Intermitten failure detected: /marvin/tests/smoke/test_primary_storage.py
Intermitten failure detected: /marvin/tests/smoke/test_privategw_acl.py
Intermitten failure detected: /marvin/tests/smoke/test_projects.py
Intermitten failure detected: /marvin/tests/smoke/test_public_ip_range.py
Intermitten failure detected: /marvin/tests/smoke/test_reset_vm_on_reboot.py
Intermitten failure detected: /marvin/tests/smoke/test_router_dhcphosts.py
Intermitten failure detected: /marvin/tests/smoke/test_router_dns.py
Intermitten failure detected: /marvin/tests/smoke/test_router_dnsservice.py
Intermitten failure detected: /marvin/tests/smoke/test_routers_iptables_default_policy.py
Intermitten failure detected: /marvin/tests/smoke/test_routers_network_ops.py
Intermitten failure detected: /marvin/tests/smoke/test_routers.py
Intermitten failure detected: /marvin/tests/smoke/test_secondary_storage.py
Intermitten failure detected: /marvin/tests/smoke/test_service_offerings.py
Intermitten failure detected: /marvin/tests/smoke/test_snapshots.py
Intermitten failure detected: /marvin/tests/smoke/test_ssvm.py
Intermitten failure detected: /marvin/tests/smoke/test_templates.py
Intermitten failure detected: /marvin/tests/smoke/test_usage.py
Intermitten failure detected: /marvin/tests/smoke/test_vm_life_cycle.py
Intermitten failure detected: /marvin/tests/smoke/test_vm_snapshots.py
Intermitten failure detected: /marvin/tests/smoke/test_volumes.py
Intermitten failure detected: /marvin/tests/smoke/test_vpc_redundant.py
Intermitten failure detected: /marvin/tests/smoke/test_vpc_router_nics.py
Intermitten failure detected: /marvin/tests/smoke/test_vpc_vpn.py
Intermitten failure detected: /marvin/tests/smoke/test_host_maintenance.py
Intermitten failure detected: /marvin/tests/smoke/test_hostha_kvm.py
Smoke tests completed. 21 look OK, 46 have error(s)
Only failed tests results shown below:


Test | Result | Time (s) | Test File
--- | --- | --- | ---
test_01_create_iso_with_checksum_sha1 | `Error` | 65.44 | test_iso.py
test_02_create_iso_with_checksum_sha256 | `Error` | 65.42 | test_iso.py
test_03_create_iso_with_checksum_md5 | `Error` | 65.42 | test_iso.py
test_04_create_iso_with_no_checksum | `Error` | 65.44 | test_iso.py
test_01_create_iso | `Failure` | 1514.09 | test_iso.py
ContextSuite context=TestISO>:setup | `Error` | 3029.01 | test_iso.py
ContextSuite context=TestAccounts>:setup | `Error` | 0.00 | test_accounts.py
ContextSuite context=TestAddVmToSubDomain>:setup | `Error` | 0.00 | test_accounts.py
test_DeleteDomain | `Error` | 1.38 | test_accounts.py
test_forceDeleteDomain | `Error` | 1.37 | test_accounts.py
ContextSuite context=TestRemoveUserFromAccount>:setup | `Error` | 5.70 | test_accounts.py
ContextSuite context=TestTemplateHierarchy>:setup | `Error` | 5.80 | test_accounts.py
ContextSuite context=TestInternalLb>:setup | `Error` | 0.00 | test_internal_lb.py
ContextSuite context=TestDeployVmWithAffinityGroup>:setup | `Error` | 0.00 | test_affinity_groups_projects.py
test_list_infrastructure_metrics | `Error` | 0.26 | test_metrics_api.py
test_list_pstorage_metrics | `Error` | 0.18 | test_metrics_api.py
test_list_vms_metrics | `Error` | 0.17 | test_metrics_api.py
ContextSuite context=TestDeployVmWithAffinityGroup>:setup | `Error` | 0.00 | test_affinity_groups.py
ContextSuite context=TestNetworkMigration>:setup | `Error` | 0.00 | test_migration.py
ContextSuite context=TestDeployVirtioSCSIVM>:setup | `Error` | 0.00 | test_deploy_virtio_scsi_vm.py
test_deploy_vm_from_iso | `Error` | 0.00 | test_deploy_vm_iso.py
ContextSuite context=TestNetworkACL>:setup | `Error` | 0.00 | test_network_acl.py
ContextSuite context=TestDeployVmRootSize>:setup | `Error` | 0.00 | test_deploy_vm_root_resize.py
ContextSuite context=TestDeployVmWithVariedPlanners>:setup | `Error` | 0.00 | test_deploy_vms_with_varied_deploymentplanners.py
test_delete_account | `Error` | 0.76 | test_network.py
test_delete_network_while_vm_on_it | `Error` | 0.08 | test_network.py
test_deploy_vm_l2network | `Error` | 0.07 | test_network.py
test_l2network_restart | `Error` | 1.18 | test_network.py
ContextSuite context=TestPortForwarding>:setup | `Error` | 3.50 | test_network.py
test_reboot_router | `Error` | 0.09 | test_network.py
test_releaseIP | `Error` | 0.76 | test_network.py
ContextSuite context=TestRouterRules>:setup | `Error` | 0.85 | test_network.py
ContextSuite context=TestDeployVmWithUserData>:setup | `Error` | 0.00 | test_deploy_vm_with_userdata.py
test_01_nic | `Error` | 0.09 | test_nic.py
ContextSuite context=TestLoadBalance>:setup | `Error` | 0.00 | test_loadbalance.py
test_UpdateStorageOverProvisioningFactor | `Failure` | 0.02 | test_over_provisioning.py
test_UpdateStorageOverProvisioningFactor | `Error` | 0.02 | test_over_provisioning.py
ContextSuite context=TestListIdsParams>:setup | `Error` | 0.00 | test_list_ids_parameter.py
test_nic_secondaryip_add_remove | `Error` | 0.08 | test_multipleips_per_nic.py
ContextSuite context=TestNestedVirtualization>:setup | `Error` | 0.00 | test_nested_virtualization.py
ContextSuite context=TestIsolatedNetworksPasswdServer>:setup | `Error` | 0.00 | test_password_server.py
ContextSuite context=TestPortForwardingRules>:setup | `Error` | 0.00 | test_portforwardingrules.py
test_01_add_primary_storage_disabled_host | `Error` | 32.61 | test_primary_storage.py
ContextSuite context=TestStorageTags>:setup | `Error` | 31.92 | test_primary_storage.py
ContextSuite context=TestPrivateGwACL>:setup | `Error` | 0.00 | test_privategw_acl.py
test_09_project_suspend | `Error` | 1.11 | test_projects.py
test_10_project_activation | `Error` | 1.08 | test_projects.py
test_02_create_template_with_checksum_sha1 | `Error` | 65.50 | test_templates.py
test_03_create_template_with_checksum_sha256 | `Error` | 65.49 | test_templates.py
test_04_create_template_with_checksum_md5 | `Error` | 65.49 | test_templates.py
test_05_create_template_with_no_checksum | `Error` | 65.48 | test_templates.py
test_02_deploy_vm_from_direct_download_template | `Error` | 0.12 | test_templates.py
test_03_deploy_vm_wrong_checksum | `Error` | 0.13 | test_templates.py
ContextSuite context=TestTemplates>:setup | `Error` | 3.53 | test_templates.py
ContextSuite context=TestResetVmOnReboot>:setup | `Error` | 0.00 | test_reset_vm_on_reboot.py
ContextSuite context=TestRouterDHCPHosts>:setup | `Error` | 0.00 | test_router_dhcphosts.py
ContextSuite context=TestRouterDHCPOpts>:setup | `Error` | 0.00 | test_router_dhcphosts.py
ContextSuite context=TestRouterDns>:setup | `Error` | 0.00 | test_router_dns.py
ContextSuite context=TestRouterDnsService>:setup | `Error` | 0.00 | test_router_dnsservice.py
ContextSuite context=TestRouterIpTablesPolicies>:setup | `Error` | 0.00 | test_routers_iptables_default_policy.py
ContextSuite context=TestVPCIpTablesPolicies>:setup | `Error` | 0.00 | test_routers_iptables_default_policy.py
test_01_isolate_network_FW_PF_default_routes_egress_true | `Error` | 0.16 | test_routers_network_ops.py
test_02_isolate_network_FW_PF_default_routes_egress_false | `Error` | 0.17 | test_routers_network_ops.py
ContextSuite context=TestRedundantIsolateNetworks>:setup | `Error` | 1.38 | test_routers_network_ops.py
ContextSuite context=TestISOUsage>:setup | `Error` | 0.00 | test_usage.py
ContextSuite context=TestLBRuleUsage>:setup | `Error` | 0.00 | test_usage.py
ContextSuite context=TestNatRuleUsage>:setup | `Error` | 0.00 | test_usage.py
ContextSuite context=TestPublicIPUsage>:setup | `Error` | 0.00 | test_usage.py
ContextSuite context=TestSnapshotUsage>:setup | `Error` | 0.00 | test_usage.py
ContextSuite context=TestVmUsage>:setup | `Error` | 0.00 | test_usage.py
ContextSuite context=TestVolumeUsage>:setup | `Error` | 0.00 | test_usage.py
ContextSuite context=TestVpnUsage>:setup | `Error` | 0.00 | test_usage.py
ContextSuite context=TestRouterServices>:setup | `Error` | 0.00 | test_routers.py
test_01_sys_vm_start | `Failure` | 0.11 | test_secondary_storage.py
test_02_sys_template_ready | `Failure` | 0.11 | test_secondary_storage.py
ContextSuite context=TestServiceOfferings>:setup | `Error` | 0.23 | test_service_offerings.py
ContextSuite context=TestSnapshotRootDisk>:setup | `Error` | 0.00 | test_snapshots.py
test_01_list_sec_storage_vm | `Failure` | 0.04 | test_ssvm.py
test_02_list_cpvm_vm | `Failure` | 0.04 | test_ssvm.py
test_03_ssvm_internals | `Failure` | 0.04 | test_ssvm.py
test_04_cpvm_internals | `Failure` | 0.04 | test_ssvm.py
test_05_stop_ssvm | `Failure` | 0.04 | test_ssvm.py
test_06_stop_cpvm | `Failure` | 0.04 | test_ssvm.py
test_07_reboot_ssvm | `Failure` | 0.04 | test_ssvm.py
test_08_reboot_cpvm | `Failure` | 0.04 | test_ssvm.py
test_09_destroy_ssvm | `Failure` | 0.04 | test_ssvm.py
test_10_destroy_cpvm | `Failure` | 0.04 | test_ssvm.py
ContextSuite context=TestDeployVM>:setup | `Error` | 0.00 | test_vm_life_cycle.py
ContextSuite context=TestVMLifeCycle>:setup | `Error` | 0.00 | test_vm_life_cycle.py
ContextSuite context=TestVmSnapshot>:setup | `Error` | 0.08 | test_vm_snapshots.py
ContextSuite context=TestCreateVolume>:setup | `Error` | 0.00 | test_volumes.py
ContextSuite context=TestVolumes>:setup | `Error` | 0.00 | test_volumes.py
test_01_create_redundant_VPC_2tiers_4VMs_4IPs_4PF_ACL | `Error` | 9.14 | test_vpc_redundant.py
test_02_redundant_VPC_default_routes | `Error` | 8.10 | test_vpc_redundant.py
test_03_create_redundant_VPC_1tier_2VMs_2IPs_2PF_ACL_reboot_routers | `Error` | 7.09 | test_vpc_redundant.py
test_04_rvpc_network_garbage_collector_nics | `Error` | 8.13 | test_vpc_redundant.py
test_05_rvpc_multi_tiers | `Error` | 6.08 | test_vpc_redundant.py
test_01_VPC_nics_after_destroy | `Error` | 6.03 | test_vpc_router_nics.py
test_02_VPC_default_routes | `Error` | 6.09 | test_vpc_router_nics.py
test_01_redundant_vpc_site2site_vpn | `Failure` | 6.47 | test_vpc_vpn.py
test_01_vpc_site2site_vpn_multiple_options | `Failure` | 5.44 | test_vpc_vpn.py
test_01_vpc_remote_access_vpn | `Failure` | 4.26 | test_vpc_vpn.py
test_01_vpc_site2site_vpn | `Failure` | 4.38 | test_vpc_vpn.py
test_02_cancel_host_maintenace_with_migration_jobs | `Error` | 0.32 | test_host_maintenance.py
test_hostha_enable_ha_when_host_in_maintenance | `Error` | 3.71 | test_hostha_kvm.py

blueorangutan commented:
<b>Trillian test result (tid-2239)</b>
Environment: kvm-centos7 (x2), Advanced Networking with Mgmt server 7
Total time taken: 29590 seconds
Marvin logs: https://github.com/blueorangutan/acs-prs/releases/download/trillian/pr2309-t2239-kvm-centos7.zip
Intermitten failure detected: /marvin/tests/smoke/test_privategw_acl.py
Intermitten failure detected: /marvin/tests/smoke/test_vpc_redundant.py
Smoke tests completed. 66 look OK, 1 have error(s)
Only failed tests results shown below:


Test | Result | Time (s) | Test File
--- | --- | --- | ---
test_04_rvpc_network_garbage_collector_nics | `Failure` | 293.89 | test_vpc_redundant.py

DaanHoogland commented:
just want to see jenkins succeed

rhtyd commented:
Will open a new PR, closing this one.

rhtyd commented:
New PR https://github.com/apache/cloudstack/pull/2469/files
